### PR TITLE
refactor(rust): avoid using non-borrowing cbor fields

### PR DESCRIPTION
**Pull request by neil2468**
## Proposed Changes

Refactor `ockam_command` & `ockam_api` so that CBOR structs avoid using non-borrowing fields and make use of `CowStr` and `CowBytes`. This is all in aid of making the code base more consistent and, hopefully, memory efficient.

Also added `impl From<&String> for CowStr` to help reduce repeated code.

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

### CODE REVIEW

File: implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs [hunk 1]
```diff
@@ -16,15 +16,15 @@ pub struct CreateForwarder<'a> {
     #[cfg(feature = "tag")]
     #[n(0)] tag: TypeTag<3386455>,
     /// Address to create forwarder at.
-    #[n(1)] address: MultiAddr,
+    #[b(1)] address: MultiAddr,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs [hunk 2]
```diff
@@ -16,15 +16,15 @@ pub struct CreateForwarder<'a> {
...
     /// Forwarder alias.
     #[b(2)] alias: Option<CowStr<'a>>,
     /// Forwarding service is at rust node.
     #[n(3)] at_rust_node: bool,
     /// An authorised identity for secure channels.
     /// Only set for non-project addresses as for projects the project's
     /// authorised identity will be used.
-    #[n(4)] authorized: Option<IdentityIdentifier>
+    #[b(4)] authorized: Option<IdentityIdentifier>
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs [hunk 3]
```diff
@@ -10,7 +10,7 @@ use ockam_core::TypeTag;
 pub struct Policy {
     #[cfg(feature = "tag")]
     #[n(0)] tag: TypeTag<2000111>,
-    #[n(1)] expression: Expr,
+    #[b(1)] expression: Expr,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs [hunk 4]
```diff
@@ -33,7 +33,7 @@ impl Policy {
 pub struct PolicyList {
     #[cfg(feature = "tag")]
     #[n(0)] tag: TypeTag<3521457>,
-    #[n(1)] expressions: Vec<(Action, Expr)>,
+    #[b(1)] expressions: Vec<(Action, Expr)>,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs [hunk 5]
```diff
@@ -23,15 +23,15 @@ pub struct CreateInlet<'a> {
     /// The peer address.
     /// This can either be the address of an already
     /// created outlet, or a forwarding mechanism via ockam cloud.
-    #[n(2)] outlet_addr: MultiAddr,
+    #[b(2)] outlet_addr: MultiAddr,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs [hunk 6]
```diff
@@ -23,15 +23,15 @@ pub struct CreateInlet<'a> {
...
     /// A human-friendly alias for this portal endpoint
     #[b(3)] alias: Option<CowStr<'a>>,
     /// Enable credentials authorization
     #[n(4)] check_credential: bool,
     /// An authorised identity for secure channels.
     /// Only set for non-project addresses as for projects the project's
     /// authorised identity will be used.
-    #[n(5)] authorized: Option<IdentityIdentifier>
+    #[b(5)] authorized: Option<IdentityIdentifier>
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs [hunk 7]
```diff
@@ -18,9 +18,9 @@ use crate::route_to_multiaddr;
 #[rustfmt::skip]
 #[cbor(index_only)]
 pub enum CredentialExchangeMode {
-    #[n(0)] None,
-    #[n(1)] Oneway,
-    #[n(2)] Mutual,
+    #[b(0)] None,
+    #[b(1)] Oneway,
+    #[b(2)] Mutual,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs [hunk 8]
```diff
@@ -32,8 +32,8 @@ pub struct CreateSecureChannelRequest<'a> {
     #[n(0)] tag: TypeTag<6300395>,
     #[b(1)] pub addr: CowStr<'a>,
     #[b(2)] pub authorized_identifiers: Option<Vec<CowStr<'a>>>,
-    #[n(3)] pub credential_exchange_mode: CredentialExchangeMode,
-    #[n(4)] pub timeout: Option<Duration>
+    #[b(3)] pub credential_exchange_mode: CredentialExchangeMode,
+    #[b(4)] pub timeout: Option<Duration>
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs [hunk 9]
```diff
@@ -13,18 +13,17 @@ use ockam_core::TypeTag;
...
     #[cfg(feature = "tag")]
-    #[n(0)]
-    tag: TypeTag<1503320>,
+    #[n(0)] tag: TypeTag<1503320>,
     /// The type of transport to create
-    #[n(1)] pub tt: TransportType,
+    #[b(1)] pub tt: TransportType,
     /// The mode the transport should operate in
-    #[n(2)] pub tm: TransportMode,
+    #[b(2)] pub tm: TransportMode,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs [hunk 10]
```diff
@@ -68,11 +66,11 @@ impl<'a> DeleteTransport<'a> {
...
 pub enum TransportType {
     /// Ockam TCP transport
-    #[n(0)] Tcp,
+    #[b(0)] Tcp,
     /// Embedded BLE transport
-    #[n(1)] Ble,
+    #[b(1)] Ble,
     /// Websocket transport
-    #[n(2)] WebSocket,
+    #[b(2)] WebSocket,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs [hunk 11]
```diff
@@ -114,20 +112,20 @@ pub struct TransportStatus<'a> {
     #[cfg(feature = "tag")]
     #[n(0)] tag: TypeTag<1581592>,
     /// The type of transport to create
-    #[n(2)] pub tt: TransportType,
+    #[b(2)] pub tt: TransportType,
     /// The mode the transport should operate in
-    #[n(3)] pub tm: TransportMode,
+    #[b(3)] pub tm: TransportMode,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: !

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs [hunk 12]
```diff
@@ -90,9 +88,9 @@ impl Display for TransportType {
 #[rustfmt::skip]
 pub enum TransportMode {
     /// Listen on a set address
-    #[n(0)] Listen,
+    #[b(0)] Listen,
     /// Connect to a remote peer
-    #[n(1)] Connect,
+    #[b(1)] Connect,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: !

**Review by adrianbenavides — COMMENTED** [review]
Hey @neil2468 thanks for submitting this PR!

As far as I know, `b` is only useful if the type has lifetimes. See [here](https://twittner.gitlab.io/minicbor/minicbor_derive/index.html#n-and-b-or-cborn-and-cborb). If the type has no lifetimes, then we should use `n`.

I've marked the places where `n` should be kept with `!`.

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs (hunk 5 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: Done 👍 

Interestingly, there was also a `listen_addr: SocketAddr` in this struct (introduced by twittner) which was borrowing. Following our logic, I changed it to non-borrowing.

File: implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs (hunk 2 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs (hunk 3 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs (hunk 4 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/policy.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs (hunk 6 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs (hunk 7 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs (hunk 8 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/secure_channel.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs (hunk 9 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs (hunk 10 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs (hunk 12 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs (hunk 11 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/transport.rs]: Done 👍

File: implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs (hunk 1 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs]: Done 👍

**Comment by neil2468** [comment]
Thanks @adrianbenavides. I've done the changes.

I have to admit, I don't 100% understand what's going on with the b/n attributes.

**Comment by adrianbenavides** [comment]
> Thanks @adrianbenavides. I've done the changes.

Great, thanks @neil2468, I'll merge the PR shortly!

> I have to admit, I don't 100% understand what's going on with the b/n attributes.

Their purpose is just to define whether a type should borrow or clone when decoding from a CBOR buffer. As a rule of thumb, If a type has lifetimes, then we should use `b` so it can be decoded from the CBOR input without allocating new memory.

**Review by adrianbenavides — APPROVED** [review]
LGTM! Thanks again @neil2468 :muscle:

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs (hunk 5 above)
Comment (adrianbenavides) [reply implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: Borrowing probably worked in this case because `SocketAddr` implements `Copy`.

**Pull request merged**